src/NetscriptFunctions.ts

```diff
@@ -1913,6 +1909,34 @@ export const ns = {
   ...NetscriptExtra(),
 };
 
+const wrappedNS = wrapAPILayer({} as ExternalAPI<NSFull>, ns, []);
+function freezeEnums(obj: object, isEnums = false) {
+  if (isEnums) Object.freeze(obj);
+  for (const [k, v] of Object.entries(obj)) {
+    if (isEnums) Object.freeze(v);
+    else if (typeof obj === "object") {
+      freezeEnums(v, k === "enums");
+    }
+  }
+}
+freezeEnums(wrappedNS);
+
+export function NetscriptFunctions(ws: WorkerScript): ExternalAPI<NSFull> {
+  const instance = new StampedLayer(ws, wrappedNS) as any;
+  instance.args = ws.args.slice();
+  stampLayers(ws, instance, wrappedNS);
+  return instance;
+}
+
+function stampLayers(ws: WorkerScript, stamped: any, unstamped: any) {
+  for (const [k, v] of Object.entries(unstamped)) {
```

Comment:
Suggestion: We've already figured out the nature of all the keys in wrapAPILayer. If that returned a secondary object that was just a pruned tree of API names, we could traverse just that and avoid checking the whole API tree at script creation time.

Reply:
Addressed in 1dc7bea